index/segment/mem/segment.go

```diff
@@ -93,26 +97,60 @@ func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 
 func (s *segment) Insert(d doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
 		return sgmt.ErrClosed
 	}
 
 	if s.state.sealed {
-		s.state.RUnlock()
 		return errSegmentSealed
 	}
 
-	// Validate that the document contains only valid UTF-8.
+	// Check if the document contains a document ID and ensure it's fields are valid UTF-8.
+	var docID []byte
```

Comment:
could you decouple the validation into a separate function that can be used upstream too?

Something like:
```
func Validate(d doc.Document) error {
  // do all `pure` Field validation (w/o touching any of the segment internals
}

func (s *segment) InsertValidated(d doc.Document) error {
  // skip any validation, only do insertion 
}

func (s *segment) Insert(d doc.Document) error {
 // do validation
 // and then insert
}